src/node/rpc/nodefrontend.h

```diff
@@ -1,95 +1,206 @@
+                this->network.secrets->get_current_version()}});
+
+          return add_node(args.tx, caller_pem_raw, in, joining_node_status);
         }
 
-#ifdef GET_QUOTE
-        QuoteVerificationResult verify_result = QuoteVerifier::verify_quote(
-          args.tx, network, in.quote, caller_pem_raw);
+        // If the service is open, new nodes are first added as pending and
+        // then only trusted via member governance. It is expected that a new
+        // node polls the network to retrieve the network secrets until it is
+        // trusted
 
-        if (verify_result != QuoteVerificationResult::VERIFIED)
-          return QuoteVerifier::quote_verification_error_to_json(verify_result);
-#else
-        LOG_INFO_FMT("Skipped joining node quote verification");
-#endif
+        auto existing_node_id = check_node_exists(args.tx, caller_pem_raw);
+        if (existing_node_id.has_value())
+        {
+          // If the node already exists, return network secrets if is already
+          // trusted. Otherwise, only return its node id
+          auto node_status = nodes_view->get(existing_node_id.value())->status;
+          if (node_status == NodeStatus::TRUSTED)
+          {
+            return jsonrpc::success<JoinNetworkNodeToNode::Out>(
+              {node_status,
+               existing_node_id.value(),
+               {this->network.secrets->get_current(),
+                this->network.secrets->get_current_version()}});
+          }
+          else if (node_status == NodeStatus::PENDING)
+          {
+            return jsonrpc::success<JoinNetworkNodeToNode::Out>(
+              {node_status, existing_node_id.value()});
+          }
+          else
+          {
+            return jsonrpc::error(
+              jsonrpc::StandardErrorCodes::INVALID_REQUEST,
+              "Joining node is not in expected state");
+          }
+        }
+        else
+        {
+          // If the node does not exist, add it to the KV in state pending
 
-        NodeId joining_node_id =
-          get_next_id(args.tx.get_view(network.values), NEXT_NODE_ID);
-
-        nodes_view->put(
-          joining_node_id,
-          {in.node_info_network,
-           caller_pem_raw,
-           in.quote,
-           NodeStatus::TRUSTED});
-
-        // Set joiner's fresh key for encrypting past network secrets
-        node.set_joiner_key(joining_node_id, args.params["raw_fresh_key"]);
-
-        LOG_INFO_FMT(
-          "Adding node {}:{} as node {}",
-          in.node_info_network.host,
-          in.node_info_network.rpcport,
-          joining_node_id);
-
-        // Send network secrets and NodeID
-        return jsonrpc::success(nlohmann::json(
-          JoinNetworkNodeToNode::Out{joining_node_id,
-                                     network.secrets->get_current(),
-                                     network.secrets->get_current_version()}));
+          // TODO: We should also automatically stage a vote for members to
```

Comment:
GitHub issue for this?

Reply:
I wanted to address this within https://github.com/microsoft/CCF/issues/293 but the ticket has been opened for too long. I've just raised https://github.com/microsoft/CCF/issues/397